pkg/types/codec.go

```diff
@@ -40,6 +40,7 @@ type TypeProvider interface {
 
 type ContractTypeProvider interface {
 	CreateContractType(contractName, itemType string, forEncoding bool) (any, error)
+	CreateContractTypeByKey(key string, forEncoding bool) (any, error)
```

Comment:
What does a "key" look like?  I feel like `CreateContractType` should just call `CreateContractTypeByKey` using some character or something to combine the contractName and itemType

Reply:
Key can be anything. For EVM its one of these:
1. Event signature:  "0x..."
2. Event signature and one of the topics "0x-value" which for `QueryKeyByValue()` could indicate that we want all USDC transfer logs and we want to filter them based on their topic value
3. Event signature and EVM word bytes eg. "0x-word=1" or something similar which for `QueryKeyByValue()`  means that we are querying for USDC transfer logs and filtering them based on their word value. 

CreateContractType isn't always tied to an event, since it is also used for contract methods. We could map these to a key also, but I feel like it adds extra complexity

Reply:
I don't follow the logic.  They need to define a contract that the event is on (contractName) and the event emitted (itemType) for us to know what they asking to encode or decode into.  How would having a different field that you filter on impact what the data is decoded into?

Reply:
contractName and itemType are derived from the key itself since the key has a mapping to an event binding [link](https://github.com/smartcontractkit/chainlink/blob/c466c600ae0d5c2b9379f513e9460e69355ababc/core/services/relay/evm/chain_reader.go#L289). This allows product to call QueryKey... by just providing the key and it still knowing how to decode the data.

Reply:
I am just saying that I don't think `CreateContractType` should call `CreateContractTypeByKey` since that would mean defining keys for methods which don't need keys in the first place

Reply:
Scratch that, I refreshed my memory on the evm impl. Yeah, what you said in the first comment makes sense